ocaml/lambda/simplif.ml

```diff
@@ -586,13 +586,12 @@ let simplify_lets lam =
           end
       | _ -> no_opt ()
       end
-  | Lfunction{kind=outer_kind; params; return=outer_return; body = l;
-              attr=attr1; loc; ret_mode; mode; region=outer_region} ->
-      begin match outer_kind, outer_region, simplif l with
-        Curried {nlocal=0},
-        true,
+  | Lfunction{kind; params; return=outer_return; body = l;
+              attr=attr1; loc; ret_mode; mode} ->
+      begin match ret_mode, simplif l with
+        Alloc_heap,
```

Comment:
This looks unsafe. For example:
```
let f (local_ _) =
  fun b -> ...
```
The new logic would compress this into a single function whose `kind` and `params` are incoherent. The old logic would prevent this because `nlocal=1` for `f`. I think checking for `nlocal=0` is good enough for this match.

Moreover, I thought we introduced n-ary lambda, so that the arity users write corresponds to the arity at runtime? Wouldn't this optimization break that?

Reply:
This was a mistake, I've asked @riaqn to fix this and push any other changes he would like.

Reply:
I've pushed commits - @mshinwell can you look at them?